File: orion-platform-service/src/services/deploy/DeployService.ts
/**
 * DeployService - Business logic layer for Deploy operations
 * 
 * Handles deployment execution, rollback, and management
 */

import pino from 'pino';

const logger = pino({ name: 'LDeploy-LService' });
import {
  DeployRepository,
  Deployment,
  DeploymentEvent,
  CreateDeploymentInput,
  UpdateDeploymentInput,
  CreateDeploymentEventInput
} from './DeployRepository';
import { getCurrentTraceId } from '../../db/tenant-context-storage';

export interface ListDeploymentsOptions {
  page?: number;
  limit?: number;
  tenantId?: string;
  projectId?: string;
  environment?: string;
  status?: string;
}

export interface PaginatedResult<T> {
  data: T[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
}

export class DeployServiceError extends Error {
  constructor(message: string, public code: string) {
    super(message);
    this.name = 'DeployServiceError';
  }
}

export class DeployService {
  private repository: DeployRepository;

  constructor(repository: DeployRepository) {
    this.repository = repository;
  }

  // ==================== Deployment CRUD ====================

  /**
   * Get deployment by ID
   */
  async getDeployment(id: string): Promise<Deployment> {
    const deployment = await this.repository.findById(id);
    
    if (!deployment) {
      throw new DeployServiceError(`Deployment not found: ${id}`, 'DEPLOY_NOT_FOUND');
    }
    
    return deployment;
  }

  /**
   * List deployments with pagination
   */
  async listDeployments(options: ListDeploymentsOptions = {}): Promise<PaginatedResult<Deployment>> {
    const { page = 1, limit = 20, tenantId, projectId, environment, status } = options;
    const offset = (page - 1) * limit;

    const [deployments, total] = await Promise.all([
      this.repository.findAll({ tenantId, projectId, environment, status, limit, offset }),
      this.repository.count({ tenantId, environment, status }),
    ]);

    return {
      data: deployments,
      total,
      page,
      limit,
      totalPages: Math.ceil(total / limit),
    };
  }

  /**
   * Create a new deployment
   */
  async createDeployment(input: CreateDeploymentInput): Promise<Deployment> {
    if (!input.tenant_id) {
      throw new DeployServiceError('Tenant ID is required', 'INVALID_INPUT');
    }

    if (!input.environment || input.environment.trim().length === 0) {
      throw new DeployServiceError('Environment is required', 'INVALID_INPUT');
    }

    const deployment = await this.repository.create({
      ...input,
      environment: input.environment.trim(),
    });

    // Log creation event
    await this.logEvent(deployment.id, 'created', 'Deployment created', input.deployed_by);

    return deployment;
  }

  /**
   * Start deployment
   */
  async startDeployment(id: string, actorId?: string): Promise<Deployment> {
    const deployment = await this.repository.findById(id);
    
    if (!deployment) {
      throw new DeployServiceError(`Deployment not found: ${id}`, 'DEPLOY_NOT_FOUND');
    }

    if (deployment.status !== 'pending') {
      throw new DeployServiceError('Can only start pending deployments', 'INVALID_STATE');
    }

    const updated = await this.repository.startDeployment(id);
    
    if (!updated) {
      throw new DeployServiceError(`Failed to start deployment: ${id}`, 'START_FAILED');
    }

    // Log start event
    await this.logEvent(id, 'started', 'Deployment started', actorId);

    // Execute deployment asynchronously
    this.executeDeployment(id).catch(err => {
      logger.error(`Deployment execution failed: ${err.message}`);
    });

    return updated;
  }

  /**
   * Execute deployment (internal method)
   */
  private async executeDeployment(deploymentId: string): Promise<void> {
    const deployment = await this.repository.findById(deploymentId);
    if (!deployment) return;

    try {
      // Simulate deployment time based on strategy
      const delay = deployment.strategy === 'blue-green' ? 1000 : 500;
      await new Promise(resolve => setTimeout(resolve, delay));

      // In real implementation, this would:
      // 1. Pull the Docker image
      // 2. Update Kubernetes deployment
      // 3. Wait for rollout
      // 4. Run health checks

      const completed = await this.repository.completeDeployment(deploymentId, 'success');
      
      if (completed) {
        await this.logEvent(deploymentId, 'completed', 'Deployment completed successfully', deployment.deployed_by || undefined);
      }
      
    } catch (error: any) {
      await this.repository.completeDeployment(deploymentId, 'failed', error.message);
      await this.logEvent(deploymentId, 'failed', `Deployment failed: ${error.message}`, deployment.deployed_by || undefined);
    }
  }

  /**
   * Cancel a deployment
   */
  async cancelDeployment(id: string, actorId?: string): Promise<Deployment> {
    const deployment = await this.repository.findById(id);
    
    if (!deployment) {
      throw new DeployServiceError(`Deployment not found: ${id}`, 'DEPLOY_NOT_FOUND');
    }

    if (deployment.status !== 'pending' && deployment.status !== 'deploying') {
      throw new DeployServiceError('Can only cancel pending or deploying deployments', 'INVALID_STATE');
    }

    const completed = await this.repository.completeDeployment(id, 'cancelled', 'Cancelled by user');
    
    if (!completed) {
      throw new DeployServiceError(`Failed to cancel deployment: ${id}`, 'CANCEL_FAILED');
    }

    await this.logEvent(id, 'cancelled', 'Deployment cancelled', actorId);

    return completed;
  }

  /**
   * Rollback to previous deployment
   */
  async rollback(id: string, actorId?: string): Promise<Deployment> {
    const current = await this.repository.findById(id);
    
    if (!current) {
      throw new DeployServiceError(`Deployment not found: ${id}`, 'DEPLOY_NOT_FOUND');
    }

    if (current.status !== 'failed' && current.status !== 'success') {
      throw new DeployServiceError('Can only rollback completed deployments', 'INVALID_STATE');
    }

    // Find previous successful deployment
    const rollbackTarget = await this.repository.findRollbackTarget(
      current.tenant_id,
      current.environment,
      current.id
    );

    if (!rollbackTarget) {
      throw new DeployServiceError('No previous deployment to rollback to', 'NO_ROLLBACK_TARGET');
    }

    // Create new deployment as rollback
    const rollback = await this.repository.create({
      tenant_id: current.tenant_id,
      project_id: current.project_id || undefined,
      pipeline_run_id: current.pipeline_run_id || undefined,
      build_id: current.build_id || undefined,
      environment: current.environment,
      strategy: current.strategy,
      config: { ...current.config, isRollback: true, rollbackFrom: id },
      deployed_by: actorId,
    });

    // Log rollback event
    await this.logEvent(id, 'rollback_started', `Rolling back to deployment ${rollbackTarget.id}`, actorId);
    await this.logEvent(rollback.id, 'rollback_target', `Rollback target from deployment ${rollbackTarget.id}`, actorId);

    // Start the rollback deployment
    await this.startDeployment(rollback.id, actorId);

    // Update original deployment with rollback reference
    await this.repository.update(id, { rollback_to: rollback.id } as UpdateDeploymentInput);

    return rollback;
  }

  // ==================== Deployment Events ====================

  /**
   * Get deployment events
   */
  async getDeploymentEvents(deploymentId: string): Promise<DeploymentEvent[]> {
    return this.repository.findEvents(deploymentId);
  }

  /**
   * Log deployment event
   */
  private async logEvent(deploymentId: string, eventType: string, message: string, actorId?: string): Promise<void> {
    try {
      await this.repository.createEvent({
        deployment_id: deploymentId,
        event_type: eventType,
        message,
        actor_id: actorId,
      });
    } catch (err) {
      logger.error('Failed to log deployment event:', err);
    }
  }

  // ==================== Queries ====================

  /**
   * Get latest deployment by environment
   */
  async getLatestDeployment(tenantId: string, environment: string): Promise<Deployment | null> {
    return this.repository.findLatestByEnvironment(tenantId, environment);
  }

  /**
   * Get deployments by build
   */
  async getDeploymentsByBuild(buildId: string): Promise<Deployment[]> {
    return this.repository.findByBuild(buildId);
  }

  /**
   * Get environments
   */
  async getEnvironments(tenantId: string): Promise<string[]> {
    return this.repository.getEnvironments(tenantId);
  }

  /**
   * Get deployment statistics
   */
  async getDeployStats(tenantId?: string): Promise<{
    total: number;
    success: number;
    failed: number;
    deploying: number;
    avgDuration: number;
  }> {
    return this.repository.getDeployStats(tenantId);
  }
}

File: blueprints/orion-ci-cd-svc-go.archived/internal/deploy/service/deploy_service.go
package service

import (
	"context"
	"errors"
	"fmt"

	"orion/ci-cd-svc-go/internal/deploy/models"
	"orion/go-common/pkg/otel"
	"orion/ci-cd-svc-go/internal/deploy/repository"

	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.uber.org/zap"
)

var (
	ErrDeploymentNotFound = errors.New("deployment not found")
	ErrInvalidStatus      = errors.New("invalid status transition")
	ErrNoRollbackTarget   = errors.New("no previous deployment found for rollback")
)

type DeployService struct {
	repo   *repository.DeploymentRepository
	logger *zap.Logger
}

func NewDeployService(repo *repository.DeploymentRepository, logger *zap.Logger) *DeployService {
	return &DeployService{repo: repo, logger: logger}
}

// ==================== CRUD ====================

// Create creates a new deployment in pending state.
func (s *DeployService) Create(ctx context.Context, d *models.Deployment) error {
	ctx, span := otel.Tracer("orion-deploy-svc").Start(ctx, "DeployService.Create")
	defer span.End()

	if d.Status == "" {
		d.Status = "pending"
	}
	if d.Strategy == "" {
		d.Strategy = "rolling"
	}

	if err := s.repo.Create(ctx, d); err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return err
	}

	// Log creation event
	_ = s.logEvent(ctx, d.ID, "created", "Deployment created", &d.DeployedBy)

	span.SetAttributes(attribute.String("deployment.id", d.ID))
	return nil
}

// GetByID retrieves a deployment by ID.
func (s *DeployService) GetByID(ctx context.Context, tenantID, id string) (*models.Deployment, error) {
	ctx, span := otel.Tracer("orion-deploy-svc").Start(ctx, "DeployService.GetByID")
	defer span.End()

	dep, err := s.repo.GetByID(ctx, tenantID, id)
	if err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return nil, ErrDeploymentNotFound
	}
	return dep, nil
}

// List returns paginated deployments.
func (s *DeployService) List(ctx context.Context, tenantID string, offset, limit int) ([]models.Deployment, error) {
	ctx, span := otel.Tracer("orion-deploy-svc").Start(ctx, "DeployService.List")
	defer span.End()

	return s.repo.List(ctx, tenantID, offset, limit)
}

// ListByFilter returns filtered, paginated deployments.
func (s *DeployService) ListByFilter(ctx context.Context, tenantID, environment, status string, offset, limit int) ([]models.Deployment, error) {
	ctx, span := otel.Tracer("orion-deploy-svc").Start(ctx, "DeployService.ListByFilter")
	defer span.End()

	return s.repo.ListByFilter(ctx, tenantID, environment, status, offset, limit)
}

// Update persists changes to a deployment.
func (s *DeployService) Update(ctx context.Context, d *models.Deployment) error {
	ctx, span := otel.Tracer("orion-deploy-svc").Start(ctx, "DeployService.Update")
	defer span.End()

	return s.repo.Update(ctx, d)
}

// Delete removes a deployment.
func (s *DeployService) Delete(ctx context.Context, tenantID, id string) error {
	ctx, span := otel.Tracer("orion-deploy-svc").Start(ctx, "DeployService.Delete")
	defer span.End()

	return s.repo.Delete(ctx, tenantID, id)
}

// Count returns total deployments for a tenant.
func (s *DeployService) Count(ctx context.Context, tenantID string) (int, error) {
	ctx, span := otel.Tracer("orion-deploy-svc").Start(ctx, "DeployService.Count")
	defer span.End()

	return s.repo.Count(ctx, tenantID)
}

// ==================== Status Transitions ====================

// StartDeployment transitions from pending to deploying.
func (s *DeployService) StartDeployment(ctx context.Context, tenantID, id string) (*models.Deployment, error) {
	ctx, span := otel.Tracer("orion-deploy-svc").Start(ctx, "DeployService.StartDeployment")
	defer span.End()

	dep, err := s.repo.GetByID(ctx, tenantID, id)
	if err != nil {
		span.RecordError(err)
		return nil, ErrDeploymentNotFound
	}

	if dep.Status != "pending" {
		err := fmt.Errorf("%w: current status is %s, expected pending", ErrInvalidStatus, dep.Status)
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return nil, err
	}

	updated, err := s.repo.StartDeployment(ctx, tenantID, id)
	if err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return nil, fmt.Errorf("failed to start deployment: %w", err)
	}

	_ = s.logEvent(ctx, id, "started", "Deployment started", &tenantID)

	span.SetAttributes(attribute.String("deployment.id", id))
	return updated, nil
}

// CompleteDeployment transitions to a terminal state (success/failed).
func (s *DeployService) CompleteDeployment(ctx context.Context, tenantID, id, status string, errorMsg *string) (*models.Deployment, error) {
	ctx, span := otel.Tracer("orion-deploy-svc").Start(ctx, "DeployService.CompleteDeployment")
	defer span.End()

	span.SetAttributes(attribute.String("target_status", status))

	dep, err := s.repo.GetByID(ctx, tenantID, id)
	if err != nil {
		span.RecordError(err)
		return nil, ErrDeploymentNotFound
	}

	if dep.Status != "deploying" {
		err := fmt.Errorf("%w: current status is %s, expected deploying", ErrInvalidStatus, dep.Status)
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return nil, err
	}

	if status != "success" && status != "failed" {
		err := fmt.Errorf("%w: target status must be success or failed, got %s", ErrInvalidStatus, status)
		span.RecordError(err)
		return nil, err
	}

	updated, err := s.repo.CompleteDeployment(ctx, tenantID, id, status, errorMsg)
	if err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return nil, fmt.Errorf("failed to complete deployment: %w", err)
	}

	eventMsg := "Deployment completed successfully"
	if status == "failed" {
		eventMsg = fmt.Sprintf("Deployment failed: %s", safeStr(errorMsg))
	}
	_ = s.logEvent(ctx, id, status, eventMsg, &tenantID)

	span.SetAttributes(attribute.String("deployment.id", id))
	return updated, nil
}

// CancelDeployment cancels a pending or deploying deployment.
func (s *DeployService) CancelDeployment(ctx context.Context, tenantID, id string) (*models.Deployment, error) {
	ctx, span := otel.Tracer("orion-deploy-svc").Start(ctx, "DeployService.CancelDeployment")
	defer span.End()

	dep, err := s.repo.GetByID(ctx, tenantID, id)
	if err != nil {
		span.RecordError(err)
		return nil, ErrDeploymentNotFound
	}

	if dep.Status != "pending" && dep.Status != "deploying" {
		err := fmt.Errorf("%w: cannot cancel deployment in %s state", ErrInvalidStatus, dep.Status)
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return nil, err
	}

	cancelMsg := "Cancelled by user"
	updated, err := s.repo.CompleteDeployment(ctx, tenantID, id, "cancelled", &cancelMsg)
	if err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return nil, fmt.Errorf("failed to cancel deployment: %w", err)
	}

	_ = s.logEvent(ctx, id, "cancelled", "Deployment cancelled", &tenantID)

	span.SetAttributes(attribute.String("deployment.id", id))
	return updated, nil
}

// ==================== Rollback ====================

// Rollback creates a new deployment that reverts to the previous successful version.
func (s *DeployService) Rollback(ctx context.Context, tenantID, id string) (*models.Deployment, error) {
	ctx, span := otel.Tracer("orion-deploy-svc").Start(ctx, "DeployService.Rollback")
	defer span.End()

	current, err := s.repo.GetByID(ctx, tenantID, id)
	if err != nil {
		span.RecordError(err)
		return nil, ErrDeploymentNotFound
	}

	if current.Status != "failed" && current.Status != "success" {
		err := fmt.Errorf("%w: can only rollback completed deployments, current status is %s", ErrInvalidStatus, current.Status)
		span.RecordError(err)
		return nil, err
	}

	// Find the previous successful deployment as rollback target
	target, err := s.repo.FindRollbackTarget(ctx, tenantID, current.Environment, current.ID)
	if err != nil {
		span.RecordError(err)
		return nil, ErrNoRollbackTarget
	}

	// Create new deployment for the rollback
	rollbackDep := &models.Deployment{
		TenantID:    tenantID,
		Environment: current.Environment,
		ServiceName: current.ServiceName,
		Version:     target.Version,
		ImageTag:    target.ImageTag,
		Status:      "pending",
		Strategy:    current.Strategy,
		DeployedBy:  current.DeployedBy,
	}

	if err := s.repo.Create(ctx, rollbackDep); err != nil {
		span.RecordError(err)
		return nil, err
	}

	// Update original deployment with rollback reference
	_ = s.repo.UpdateRollbackTo(ctx, tenantID, id, rollbackDep.ID)

	// Log events
	_ = s.logEvent(ctx, id, "rollback_started", fmt.Sprintf("Rolling back to deployment %s", target.ID), &tenantID)
	_ = s.logEvent(ctx, rollbackDep.ID, "rollback_target", fmt.Sprintf("Rollback from deployment %s", id), &tenantID)

	// Start the rollback deployment
	started, err := s.repo.StartDeployment(ctx, tenantID, rollbackDep.ID)
	if err != nil {
		s.logger.Warn("failed to auto-start rollback deployment", zap.Error(err))
	} else {
		rollbackDep = started
	}

	span.SetAttributes(
		attribute.String("deployment.id", id),
		attribute.String("rollback.id", rollbackDep.ID),
		attribute.String("rollback.target", target.ID),
	)
	return rollbackDep, nil
}

// ==================== Queries ====================

// GetLatestDeployment returns the latest deployment for a tenant+environment.
func (s *DeployService) GetLatestDeployment(ctx context.Context, tenantID, environment string) (*models.Deployment, error) {
	ctx, span := otel.Tracer("orion-deploy-svc").Start(ctx, "DeployService.GetLatestDeployment")
	defer span.End()

	dep, err := s.repo.FindLatestByEnvironment(ctx, tenantID, environment)
	if err != nil {
		return nil, err
	}
	return dep, nil
}

// GetDeploymentsByBuild returns all deployments for a build, scoped to tenant.
func (s *DeployService) GetDeploymentsByBuild(ctx context.Context, tenantID, buildID string) ([]models.Deployment, error) {
	ctx, span := otel.Tracer("orion-deploy-svc").Start(ctx, "DeployService.GetDeploymentsByBuild")
	defer span.End()

	return s.repo.FindByBuild(ctx, tenantID, buildID)
}

// GetEnvironments returns distinct environments for a tenant.
func (s *DeployService) GetEnvironments(ctx context.Context, tenantID string) ([]string, error) {
	ctx, span := otel.Tracer("orion-deploy-svc").Start(ctx, "DeployService.GetEnvironments")
	defer span.End()

	return s.repo.GetEnvironments(ctx, tenantID)
}

// GetDeployStats returns aggregate deployment statistics.
func (s *DeployService) GetDeployStats(ctx context.Context, tenantID string) (*models.DeployStats, error) {
	ctx, span := otel.Tracer("orion-deploy-svc").Start(ctx, "DeployService.GetDeployStats")
	defer span.End()

	return s.repo.GetDeployStats(ctx, tenantID)
}

// ==================== Events ====================

// GetDeploymentEvents returns all events for a deployment.
func (s *DeployService) GetDeploymentEvents(ctx context.Context, deploymentID string) ([]models.DeploymentEvent, error) {
	ctx, span := otel.Tracer("orion-deploy-svc").Start(ctx, "DeployService.GetDeploymentEvents")
	defer span.End()

	return s.repo.FindEvents(ctx, deploymentID)
}

// logEvent creates a deployment event. Errors are logged but not propagated.
func (s *DeployService) logEvent(ctx context.Context, deploymentID, eventType, message string, actorID *string) error {
	e := &models.DeploymentEvent{
		DeploymentID: deploymentID,
		EventType:    eventType,
		Message:      &message,
		ActorID:      actorID,
	}
	if err := s.repo.CreateEvent(ctx, e); err != nil {
		s.logger.Warn("failed to log deployment event",
			zap.String("deployment_id", deploymentID),
			zap.String("event_type", eventType),
			zap.Error(err),
		)
		return err
	}
	return nil
}

func safeStr(s *string) string {
	if s == nil {
		return ""
	}
	return *s
}
